radix-engine/src/object_modules/metadata/package.rs

```diff
@@ -262,8 +261,9 @@ impl MetadataNativePackage {
         Ok(Own(node_id))
     }
 
+    /// This method assumes that the data has been pre-checked.
     pub fn init_system_struct(
-        data: MetadataInit,
+        data: IndexMap<Vec<u8>, (Option<Vec<u8>>, bool)>,
```

Comment:
How confident are we that these refactors are 100% backwards-compatible across Cuttlefish in terms of protocol-visible changes? e.g. we're changing which errors we return (which I think is safe I think assuming we only error in the same cases). And e.g. we aren't making any sys-calls/billing in the middle, which seems safe?

I see we've replaced some encode unwraps with e.g. `MetadataKeyValidationError::InvalidValue` - I guess we don't actually assume this to be hit, and indeed if it unwrapped previously, it shouldn't introduce any issues?

Reply:
> How confident are we that these refactors are 100% backwards-compatible across Cuttlefish in terms of protocol-visible changes

I did think about this thoroughly. As far I can tell, there is only error code difference which should not affect consensus, but would appreciate another pair of eyes.

> I see we've replaced some encode unwraps with e.g. MetadataKeyValidationError::InvalidValue - I guess we don't actually assume this to be hit, and indeed if it unwrapped previously, it shouldn't introduce any issues?

This is right, considering `MetadataInit` often comes from an sbor-encoded inputs.

Keeping it in the validation logic so it's self-contained and no need to make assumption about sbor-able.

Reply:
Sounds great. And yes, I agree that I think it's safe, so I'm happy with this.